### DISCUSSION TITLE
How to register Inertia 1.x as a plugin during tests (Jest or Vitest)?

### DISCUSSION THREAD

**Discussion by deleugpn** [inertiajs/inertia#1683] [Help (Vue)]
I'm getting the following error: 

```
TypeError: Cannot read properties of undefined (reading 'createProvider')
    at Proxy.data (/node_modules/@inertiajs/vue3/src/head.ts:16:35)
    at applyOptions (/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:3354:30)
    at finishComponentSetup (/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7295:5)
    at setupStatefulComponent (/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7219:5)
    at setupComponent (/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7151:36)
```

This started happening because I put `<Head>` on my Vue component:

```
<Head>
    <component head-key="freshworks-asset" is="script" src="https://widget.freshworks.com/widgets/5000000215.js" async defer />
</Head>
```

This is my test:

```
const dummy = {template: '<div>dummy</div>'};

const wrapper = mount(component, {
    props: props,
    attachTo: document.body,
    global: {
      components: {
        'font-awesome-icon': FontAwesomeIcon,
      },
      plugins: [createPinia()],
    },
  });

expect(wrapper.testId(`layout/left-menu/${element}`).attributes().href).toBe(link);
```

Most content that I found online refers to Inertia 0.x, where it was possible to do:

```
import { plugin } from '@inertiajs/inertia-vue3'
```

But as we can see on the upgrade guide, this is now gone: https://inertiajs.com/upgrade-guide#renamed-imports

So, to my original question: how can I register inertia as a plugin the same way I register `pinia` or `font-awesome` in Vitest/Jest in Inertia 1.x?

**Comment by malssid** [comment]
Have you found a solution to this? Amazing how it's been 5 months and nothing from the developers.

**Reply by RobertBoes** [reply]
> Amazing how it's been 5 months and nothing from the developers.

Honestly, comments like these make contributing to an OSS package less and less fun and feeling like an actual burden, as if everything _you_ want is up to _others_ to fix. Maybe the maintainers don't use Vitepress/Jest (which is kinda obvious, as testing within Inertia itself is done with cypress, at least for the Vue 2 adapter; https://github.com/inertiajs/inertia/tree/master/packages/vue2/tests), so it would make sense that it hasn't been investigated, since there's not a lot of gain for the maintainers while needing to maintain it.

**Reply by deleugpn** [reply]
@malssid I just stopped trying to use `Head` altogether. Haven't faced this issue again since then.

@RobertBoes I get where he is coming from, Inertia is a 1st-party package on the Laravel ecosystem, but it doesn't get enough attention as it's mostly "feature-complete" but still lacks a lot of the high quality work that goes into the Laravel brand. However, I do perceive this not only as a core developers issue, but rather as a community issue. The adoption of Inertia seems to be growing, but the amount of community content is still very small. 

We're still lacking content creators that talk about the high quality Laravel content around automation testing, components best practice designed for Inertia, Typescript, IDE tips, etc. I take some of that blame myself because it's been 5 months and now I am in position of writing a little bit more about Inertia, but I've been so swamped at work because of my decision to use Inertia that I'm still buried in a slow development cycle that I need to catch up before I can take some time to produce content.

**Reply by malssid** [reply]
Hey @RobertBoes, I didn't mean to come off rude. I agree with what @deleugpn said about the missing pieces being a community issue rather than a core developers issue. We've been using Inertia over here for about 5 months and I'll run into issues where the Inertia documentation doesn't explain how to do something, so I'm stuck with stack overflow, chatgpt, laracasts, etc. For example, mocking Inertia components wasn't clear (I've figured it out) and there doesn't seem to be a way to register Inertia as a plugin globally.

**Reply by ncphillips** [reply]
Has anyone had any progress this?

As far as I can tell, the design of Inertia makes testing using Vitest (the official Vue testing library) impossible.

> Honestly, comments like these make contributing to an OSS package less and less fun and feeling like an actual burden, as if everything you want is up to others to fix.

To be honest, I disagree. This isn't a minor issue that's making it hard. Digging around the source of both Inertia and `@vue/test-utils`, I don't see how it's possible, and I don't think a PR from me completely re-archetecting the library would be well received. `@vue/test-utils` and `vitest` are the officially recommended ways to test Vue components. IMO that should be considered table-stakes when building a library like this. Especially one that's on the front-page of laravel.com
